internal/cmd/gocdk/add_test.go

```diff
@@ -0,0 +1,195 @@
+	cmd, err := alloc.start(ctx, pctx, logger, pctx.workdir)
+	if err != nil {
+		t.Fatalf("failed to start server: %v", err)
+	}
+	defer endServerProcess(cmd)
+
+	tests := []struct {
+		pt            string
+		description   string
+		urlPaths      []string
+		op            string
+		urlValues     url.Values // only used if op=POST
+		wantStatus    int
+		stringsToFind []string
+	}{
+		{
+			pt:          "blob.Bucket",
+			description: "base",
+			urlPaths: []string{
+				"/demo/blob.bucket",
+				"/demo/blob.bucket/",
+			},
+			op:         "GET",
+			wantStatus: http.StatusOK,
+			stringsToFind: []string{
+				"<title>blob.Bucket demo</title>",
+				"This page demonstrates the use of a Go CDK blob.Bucket",
+				`<a href="./list">List</a>`,
+				`<a href="./view">View</a>`,
+				`<a href="./write">Write</a>`,
+			},
+		},
+		{
+			pt:          "blob.Bucket",
+			description: "list: empty",
+			urlPaths:    []string{"/demo/blob.bucket/list"},
+			op:          "GET",
+			wantStatus:  http.StatusOK,
+			stringsToFind: []string{
+				"<title>blob.Bucket demo</title>",
+				"no blobs in bucket",
+			},
+		},
+		{
+			pt:          "blob.Bucket",
+			description: "view: missing key param",
+			urlPaths:    []string{"/demo/blob.bucket/view"},
+			op:          "GET",
+			wantStatus:  http.StatusBadRequest,
+		},
+		// TODO(rvangent): Add tests as functionality is added, including:
+		// -- /write shows a form
+		// -- /write POST creates a blob (top level, plus one in subdir)
+		// -- /list at the top level shows the subdirectory, with links
+		// -- /list of the subdirectory shows the blob
+		// -- /view missing key param shows a form with a dropdown
+		// -- /view of the blob works
+	}
+
+	for _, test := range tests {
+		t.Run(test.pt+":"+test.description, func(t *testing.T) {
+			for _, urlPath := range test.urlPaths {
+				u := alloc.url(urlPath).String()
+				var resp *http.Response
+				var err error
+				switch test.op {
+				case "GET":
+					resp, err = http.DefaultClient.Get(u)
+				case "POST":
+					resp, err = http.DefaultClient.PostForm(u, test.urlValues)
+				default:
+					t.Fatalf("invalid test.op: %q", test.op)
+				}
+				if err != nil {
+					t.Fatalf("HTTP %q request failed: %v", test.op, err)
+				}
+				defer resp.Body.Close()
+				if resp.StatusCode != test.wantStatus {
+					t.Fatalf("HTTP request returned status code %v, want %v", resp.StatusCode, test.wantStatus)
+				}
+				bodyBytes, err := ioutil.ReadAll(resp.Body)
+				if err != nil {
+					t.Errorf("failed to read HTTP response body: %v", err)
+				}
+				body := string(bodyBytes)
+				logBody := false // only log the body on failure, and only once per test
+				for _, s := range test.stringsToFind {
+					if !strings.Contains(body, s) {
+						t.Errorf("didn't find %q in HTTP response body", s)
+						logBody = true
+					}
+				}
+				if logBody {
+					t.Error("Full HTTP response body:\n\n", body)
+				}
+			}
+		})
+	}
+}
+
+func freePort() int {
```

Comment:
nit: add "find" because "free" can be ambiguous as a verb:

```suggestion
func findFreePort() int {
```

Reply:
Done.